voicevox_engine/synthesis_engine/synthesis_engine.py

```diff
@@ -476,3 +487,121 @@ def _synthesis_impl(self, query: AudioQuery, speaker_id: int):
             wave = numpy.array([wave, wave]).T
 
         return wave
+
+    def guided_synthesis(
+        self,
+        query: AudioQuery,
+        speaker: int,
+        audio_file: IO,
+        normalize: int,
+    ):
+        f0, phonemes = extract_guided_feature(audio_file, query.kana)
+
+        phone_list = numpy.zeros((len(f0), OjtPhoneme.num_phoneme), dtype=numpy.float32)
+
+        for s, e, p in phonemes:
+            s, e = (resample_ts(v) for v in (s, e))
+            if p == "silB":
+                f0[:e] = 0.0
+                s += 1
+                p = "pau"
+            elif p == "silE":
+                f0[s:] = 0.0
+                p = "pau"
+            elif p == "sp":
+                f0[s:e] = 0.0
+                p = "pau"
+            elif p == "q":
+                p = "cl"
+            phone_list[s - 1 : e] = OjtPhoneme(start=s, end=e, phoneme=p).onehot
+
+        if normalize:
+            f0 += get_normalize_diff(
+                engine=self, kana=query.kana, f0=f0, speaker_id=speaker
+            )
+
+        f0 *= 2 ** query.pitchScale
+        f0[f0 > 6.5] = 6.5
+        f0[(0 < f0) & (f0 < 3)] = 3.0
+
+        f0 = resample(f0, int(len(f0) / query.speedScale))
+        phone_list = resample(phone_list, int(len(phone_list) / query.speedScale))
+
+        wave = self.decode_forwarder(
+            length=phone_list.shape[0],
+            phoneme_size=phone_list.shape[1],
+            f0=f0[:, numpy.newaxis].astype(numpy.float32),
+            phoneme=phone_list,
+            speaker_id=numpy.array([speaker], dtype=numpy.int64).reshape(-1),
+        )
+
+        if query.volumeScale != 1:
+            wave *= query.volumeScale
+
+        if query.outputSamplingRate != self.default_sampling_rate:
+            wave = resample(
+                wave,
+                query.outputSamplingRate * len(wave) // self.default_sampling_rate,
+            )
+
+        if query.outputStereo:
+            wave = numpy.array([wave, wave]).T
+
+        return wave
+
+    def guided_accent_phrases(
```

Comment:
`guided_synthesis`・`guided_accent_phrases`・`extract_guided_feature`はそれぞれ何をしていますか？
`extract_guided_feature`はF0と音素のアライメントをしている･･･？

Reply:
`guided_synthesis` extracts f0 and alignment then pass them to `decode_forwarder`, returns the synthesized audio.

`guided_accent_phrases` does the same thing except that it returns a list of `AccentPhrase` based on f0 and alignment.

The major difference between them is `guided_synthesis` passes the f0 information in the resolution of frames while `guided_accent_phrases` processes in the resolution of phonemes. While knowing passing by frame may make no sense since it's not how `decode_forwarder` works, I have a hope that it would contain more pitch information.

> `extract_guided_feature`はF0と音素のアライメントをしている･･･？

Exactly. It just passes the raw f0 information extracted by `harvest` and the alignment text produced by `julius4seg`

Reply:
> The major difference between them is guided_synthesis passes the f0 information in the resolution of frames while guided_accent_phrases processes in the resolution of phonemes.

なるほどです！！
どれも有用な面白い機能だと思います。

しかし、この辺りは専門知識がないととても難しいはずです。
コードを読んだ開発者がなんとなく理解できるように、一言で良いのでコメントを書いておくのはどうでしょう。

---

I see!
I think they are all useful and interesting features.

However, these should be very difficult to understand without expertise.
How about writing a comment just a few words so that the developer who reads the code can somewhat understand it?